portfolio/src/main/java/com/google/sps/servlets/TeaVoteServlet.java

```diff
@@ -32,32 +33,69 @@
 
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        //Update or add flavor entity.
         String flavor = request.getParameter("flavor");
+        Query flavorQuery = new Query("FlavorVote").setFilter(
+            new Query.FilterPredicate("flavor", Query.FilterOperator.EQUAL, flavor));
+        Entity flavorVoteEntity = datastore.prepare(flavorQuery).asSingleEntity();
+        long totalFlavorVotes = 0;
+
+        if (flavorVoteEntity != null) {
+            totalFlavorVotes = (long) flavorVoteEntity.getProperty("votes");
+        } else {
+            //Flavor entity does not exist. Create a new one.
+            flavorVoteEntity = new Entity("FlavorVote");
+        }
+
+        flavorVoteEntity.setProperty("flavor", flavor);
+        flavorVoteEntity.setProperty("votes", totalFlavorVotes + 1);
+
+        //Update or add topping entity.
         String topping = request.getParameter("topping");
+        Query toppingQuery = new Query("ToppingVote").setFilter(
+            new Query.FilterPredicate("topping", Query.FilterOperator.EQUAL, topping));
+        Entity toppingVoteEntity = datastore.prepare(toppingQuery).asSingleEntity();
+        long totalToppingVotes = 0;
+
+         if (toppingVoteEntity != null) {
+            totalToppingVotes = (long) toppingVoteEntity.getProperty("votes");
+        } else {
+            //Topping entity does not exist. Create a new one.
+            toppingVoteEntity = new Entity("ToppingVote");
+        }
 
-        Entity voteEntity = new Entity("TeaVote");
-        voteEntity.setProperty("flavor", flavor);
-        voteEntity.setProperty("topping", topping);
+        toppingVoteEntity.setProperty("topping", topping);
+        toppingVoteEntity.setProperty("votes", totalToppingVotes + 1);
 
-        datastore.put(voteEntity);
+        datastore.put(flavorVoteEntity);
+        datastore.put(toppingVoteEntity);
         response.sendRedirect("/index.html");
     }
 
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        Query query = new Query("TeaVote");
-        Iterable<Entity> results = datastore.prepare(query).asIterable();
-        
-        HashMap<String, Integer> flavorVotes = new HashMap<String, Integer>();
-        HashMap<String, Integer> toppingVotes = new HashMap<String, Integer>();
+        //Creating hashmap of flavors and votes.
+        Iterable<Entity> flavorResults = datastore.prepare(new Query("FlavorVote")).asIterable();
```

Comment:
very minor nit, but try to locate your variables somewhat according to their use. the iterable here for instance, is only used in the following loop - i would locate it as close to that loop as possible. the hashmap is used in the rest of the function, so it makes sense to declare it first (aka, i would swap the line order here, init the hashmap first, then the iterable). this has no effect on the correctness of the code obviously, but makes it easier to understand that the iterable is a temporary variable created to ease readability, and the hashmap is being used longer term in the function.

Reply:
Switched. I will keep this in mind for the future!